command_line/fp3.py

```diff
@@ -0,0 +1,578 @@
+from dials_scratch.fp3 import even_blocks, index_blocks, format_phil_include
+from dials_scratch.fp3 import nproc, combine_reflections, combine_experiments
+from dials_scratch.fp3 import find_setup_script
+
+scope = phil.parse(
+    """
+find_spots {
+  include scope dials.command_line.find_spots.phil_scope
+}
+index {
+  include scope dials.command_line.index.phil_scope
+}
+refine {
+  include scope dials.command_line.refine.phil_scope
+}
+integrate {
+  include scope dials.command_line.integrate.phil_scope
+}
+symmetry {
+  include scope dials.command_line.symmetry.phil_scope
+}
+scale {
+  include scope dials.command_line.scale.phil_scope
+}
+resolution {
+  include scope dials.command_line.estimate_resolution.phil_scope
+}
+parallelism = *process drmaa
+  .type = choice
+max_workers = 1
+  .type = int
+  .help = "Maximum number of concurrent workers"
+worker_nproc = 1
+  .type = int
+  .help = "Number of processes for each worker"
+log_level = CRITICAL ERROR WARNING *INFO DEBUG
+  .type = choice
+""",
+    process_includes=True,
+)
+
+
+class FP3:
+    def __init__(self, filenames, params):
+        self._experiment = ExperimentListFactory.from_filenames(filenames)
+
+        # parse PHIL parameters
+        clai = scope.command_line_argument_interpreter()
+        self._working = scope.fetch(clai.process_and_fetch(params))
+        self._params = self._working.extract()
+        self._debug = self._params.log_level == "DEBUG"
+
+        # configure logging
+        logging.basicConfig(
+            filename="dials.fp3.log", filemode="w", format="%(message)s"
+        )
+        logger.addHandler(logging.StreamHandler(sys.stdout))
+        logger.setLevel(getattr(logging, self._params.log_level))
+        logger.info(scope.fetch_diff(self._working).as_str())
+
+        self._crystal = None
+        self._root = os.getcwd()
+        self._n = nproc()
+
+        # quick checks...
+        scan = self._experiment[0].scan
+        osc = scan.get_oscillation()
+        rng = scan.get_image_range()
+
+        self._osc = osc
+        self._rng = rng
+
+        self._integrated = []
+        self._combined = None
+        self._symmetry = None
+        self._scaled = None
+
+    def _write_phil(self, what, where):
+        """helper function: extract (what) from global scope and write as
+        (where)/(what).phil iff it contains anything. If file written,
+        returns [filename] else [] to allow straight command-line adding."""
+
+        phil = format_phil_include(scope, self._working, what).strip()
+        if not phil:
+            return []
+
+        out = os.path.join(where, f"{what}.phil")
+        with open(out, "w") as f:
+            f.write(phil)
+        return [out]
+
+    def index(self):
+        """Initial find spots and indexing: if it has been run already will
+        just reload the results from the previous run."""
+
+        work = os.path.join(self._root, "index")
+        if os.path.exists(work):
```

Comment:
```suggestion
        if os.path.exists(os.path.join(work, "indexed.expt")):
```
to avoid stacktrace if `index/` exists, but `indexed.expt` was never written to